Apps/FlightSoftware/Watchdog2/src/i2c.c

```diff
@@ -0,0 +1,318 @@
+
+    UCB0CTLW0 |= UCTR + UCTXSTT;             // I2C TX, start condition
+
+    __bis_SR_register(GIE);              // Enable interrupts
+    __delay_cycles(10000);               // give fuel gauge time to respond
+    return MasterMode;
+
+}
+
+/* For slave device with dev_addr, writes the data specified in *reg_data
+ *
+ * dev_addr: The slave device address.
+ *           Example: I2C_SLAVE_ADDR
+ * reg_addr: The register or command to send to the slave.
+ *           Example: CMD_TYPE_0_MASTER
+ * *reg_data: The buffer to write
+ *           Example: MasterType0
+ * count: The length of *reg_data
+ *           Example: TYPE_0_LENGTH
+ *  */
+I2C_Mode I2C_Master_WriteReg(uint8_t dev_addr, uint8_t reg_addr, uint8_t *reg_data, uint8_t count)
+{
+    /* Initialize state machine */
+    MasterMode = TX_REG_ADDRESS_MODE;
+    TransmitRegAddr = reg_addr;
+
+    //Copy register data to TransmitBuffer
+    CopyArray(reg_data, TransmitBuffer, count);
+
+    TXByteCtr = count;
+    RXByteCtr = 0;
+    ReceiveIndex = 0;
+    TransmitIndex = 0;
+
+    /* Initialize slave address and interrupts */
+    UCB0I2CSA = dev_addr;
+    UCB0IFG &= ~(UCTXIFG + UCRXIFG);       // Clear any pending interrupts
+    UCB0IE &= ~UCRXIE;                       // Disable RX interrupt
+    UCB0IE |= UCTXIE;                        // Enable TX interrupt
+
+    UCB0CTLW0 |= UCTR + UCTXSTT;             // I2C TX, start condition
+
+    __bis_SR_register(GIE);              // Enable interrupts
+    __delay_cycles(10000);               // give fuel gauge some time to respond
+    return MasterMode;
+}
+
+void CopyArray(uint8_t *source, uint8_t *dest, uint8_t count)
+{
+    uint8_t copyIndex = 0;
+    for (copyIndex = 0; copyIndex < count; copyIndex++)
+    {
+        dest[copyIndex] = source[copyIndex];
+    }
+}
+
+void updateGaugeReadings(){
+    // record new measurements in fuel gauge
+    readBatteryCharge();
+    readBatteryVoltage();
+    readBatteryCurrent();
+    readGaugeTemp(); //TODO: probably don't need this one
+}
+
+void fuelGaugeLowPower(){
+    // shut off all analog parts of fuel gauge circuit by setting LSB of control register to 1
+    // set 2 MSB to 00 to put in sleep mode
+    fuel_gauge_write_control_reg = 0b00101001;
+    I2C_Master_WriteReg(I2C_SLAVE_ADDR, CONTROL, &fuel_gauge_write_control_reg, 1);
+}
+
+void readBatteryCharge(){
+    I2C_Master_ReadReg(I2C_SLAVE_ADDR, ACCUMULATED_CHARGE_LSB, I2C_RX_BUFFER_MAX_SIZE);
+    CopyArray((uint8_t*)ReceiveBuffer, (uint8_t*)&raw_battery_charge[1], I2C_RX_BUFFER_MAX_SIZE);
+    I2C_Master_ReadReg(I2C_SLAVE_ADDR, ACCUMULATED_CHARGE_MSB, I2C_RX_BUFFER_MAX_SIZE);
+    CopyArray((uint8_t*)ReceiveBuffer, (uint8_t*)&raw_battery_charge, I2C_RX_BUFFER_MAX_SIZE);
+
+    // scale battery charge to fill most of 7 bit range available for telemetry
+    batt_charge_telem = (uint8_t)( (uint16_t)(raw_battery_charge[1] + (raw_battery_charge[0] << 8)) >> 10) * 3;
```

Comment:
Nit: There's a lot -- arguably too much -- happening in this single line of code. My preference would be to either split this up into several steps with informative variable names, or explain what this is doing in the comment. The current comment says what this is supposedly accomplishing, but I think this code is a bit too complicated to go without an explanation of how exactly this code accomplishes what the comment says it does.

Reply:
Wait, it doesn't even look like this is used anywhere